lib/dal/src/func/before.rs

```diff
@@ -1,55 +1,165 @@
-use serde::Deserialize;
-use veritech_client::{encrypt_value_tree, BeforeFunction};
+use thiserror::Error;
+use veritech_client::{encrypt_value_tree, BeforeFunction, CycloneValueEncryptError};
 
+use crate::attribute::value::AttributeValueError;
+use crate::prop::{PropError, PropPath};
+use crate::schema::variant::root_prop::RootPropChild;
+use crate::schema::variant::SchemaVariantError;
 use crate::{
-    standard_model, ComponentId, DalContext, EncryptedSecret, Func, FuncError, FuncResult,
+    AttributeValue, Component, ComponentError, ComponentId, DalContext, EncryptedSecret, Func,
+    FuncId, Prop, PropId, SchemaVariant, SecretError, SecretId, StandardModel, StandardModelError,
 };
 
-const AUTH_FUNCS_FOR_COMPONENT: &str =
-    include_str!("../queries/func/authentication_funcs_for_component.sql");
-
-#[derive(Deserialize, Debug)]
-struct EncryptedSecretAndFunc {
-    encrypted_secret: EncryptedSecret,
-    func: Func,
+#[remain::sorted]
+#[derive(Error, Debug)]
+pub enum BeforeFuncError {
+    #[error("attribute value error: {0}")]
+    AttributeValue(#[from] AttributeValueError),
+    #[error("component error: {0}")]
+    Component(#[from] ComponentError),
+    #[error("cyclone value encrypt error: {0}")]
+    CycloneValueEncrypt(#[from] CycloneValueEncryptError),
+    #[error("func error: {0}")]
+    Func(String),
+    #[error("error deserializing json")]
+    JsonDeserialize(#[from] serde_json::Error),
+    #[error("Function missing expected code: {0}")]
+    MissingCode(FuncId),
+    #[error("Function missing expected handler: {0}")]
+    MissingHandler(FuncId),
+    #[error("no widget options on secret prop id: {0}")]
+    NoWidgetOptionsOnSecretProp(PropId),
+    #[error("prop error: {0}")]
+    Prop(#[from] PropError),
+    #[error("schema variant error: {0}")]
+    SchemaVariant(#[from] SchemaVariantError),
+    #[error("secret error: {0}")]
+    Secret(#[from] SecretError),
+    #[error("standard model error: {0}")]
+    StandardModel(#[from] StandardModelError),
 }
 
+pub type BeforeFuncResult<T> = Result<T, BeforeFuncError>;
+
 pub async fn before_funcs_for_component(
```

Comment:
For reviewers: we acknowledge there are some algorithmic-ally "bleh" traversal methods in here. However, this was more-or-less a direct port of how `before_funcs` assembly works on `main`, so we can optimize in the future as `new-engine` becomes the new `main`.